### PR TITLE
[Fix] Fix deform conv by adding an extra argument im2col_step

### PR DESCRIPTION
**Pull request by teamwong111**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Related Issue: https://github.com/open-mmlab/mmcv/issues/1440
Add argument `im2col_step` to make it more flexible.

## Modification

Add argument `im2col_step` and modify relevant unit test

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: tests/test_ops/test_deform_conv.py [hunk 1]
```diff
@@ -34,6 +34,18 @@
 gt_deform_weight_grad = [[[[3.62, 0.], [0.40, 0.18]]]]
 
 
+def cat2batch(batch_size=1, *args):
```
Comment (zhouzaida) [inline tests/test_ops/test_deform_conv.py:37]: We can simply use `np.repeat` to expand the dim of batch so this function can be left out.

```python
np.repeat(input, batch_size)
```

File: tests/test_ops/test_deform_conv.py [hunk 2]
```diff
@@ -119,12 +152,14 @@ def _test_amp_deformconv(self, input_dtype, threshold=1e-3):
         assert np.allclose(out.data.detach().cpu().numpy(), gt_out, threshold)
         assert np.allclose(x.grad.detach().cpu().numpy(), gt_x_grad, threshold)
         assert np.allclose(
-            model.conv_offset.weight.grad.detach().cpu().numpy(),
+            model.conv_offset.weight.grad.detach().cpu().numpy() / batch_size,
```
Comment (zhouzaida) [inline tests/test_ops/test_deform_conv.py:155]: It is recommended to add a comment explaining why we divide by `batch_size` .
```
# the batch size of the input is increased which results in a larger gradient so we need to divide by the batch_size
```

File: tests/test_ops/test_deform_conv.py [hunk 3]
```diff
@@ -45,9 +45,19 @@ def _test_deformconv(self,
         from mmcv.ops import DeformConv2dPack
         c_in = 1
         c_out = 1
-        x = torch.tensor(input, device=device, dtype=dtype)
+        batch_size = 80
```
Comment (zhouzaida) [inline tests/test_ops/test_deform_conv.py:48]: We can replace `batch_size = 80` with `batch_size = 10` to reduce the memory use when running the unit test.

File: tests/test_ops/test_deform_conv.py [hunk 4]
```diff
@@ -45,9 +45,19 @@ def _test_deformconv(self,
...
+        _input = np.repeat(input, batch_size, axis=0)
+        _gt_out = np.repeat(gt_out, batch_size, axis=0)
+        _gt_x_grad = np.repeat(gt_x_grad, batch_size, axis=0)
```
Comment (zhouzaida) [inline tests/test_ops/test_deform_conv.py:51]:
```suggestion
        repeated_input = np.repeat(input, batch_size, axis=0)
        repeated_gt_out = np.repeat(gt_out, batch_size, axis=0)
        repeated_gt_x_grad = np.repeat(gt_x_grad, batch_size, axis=0)
```

File: tests/test_ops/test_deform_conv.py [hunk 5]
```diff
@@ -155,3 +184,10 @@ def test_deformconv(self):
             with autocast(enabled=True):
                 self._test_amp_deformconv(torch.float, 1e-1)
                 self._test_amp_deformconv(torch.half, 1e-1)
+
+
+# if __name__ == "__main__":
+#     _gt_out = np.repeat(gt_out, 80, axis=0)
+#     print(_gt_out.shape)
+# #     testdc = TestDeformconv()
+# #     testdc.test_deformconv()
```
Comment (zhouzaida) [inline tests/test_ops/test_deform_conv.py:193]: Those lines can be removed.

File: tests/test_ops/test_deform_conv.py [hunk 6]
```diff
@@ -102,9 +118,19 @@ def _test_amp_deformconv(self, input_dtype, threshold=1e-3):
         from mmcv.ops import DeformConv2dPack
         c_in = 1
         c_out = 1
-        x = torch.Tensor(input).cuda().type(input_dtype)
+        batch_size = 80
+        _input = np.repeat(input, batch_size, axis=0)
+        _gt_out = np.repeat(gt_out, batch_size, axis=0)
+        _gt_x_grad = np.repeat(gt_x_grad, batch_size, axis=0)
```
Comment (zhouzaida) [inline tests/test_ops/test_deform_conv.py:124]:
```suggestion
        repeated_input = np.repeat(input, batch_size, axis=0)
        repeated_gt_out = np.repeat(gt_out, batch_size, axis=0)
        repeated_gt_x_grad = np.repeat(gt_x_grad, batch_size, axis=0)
```

File: mmcv/ops/deform_conv.py [hunk 7]
```diff
@@ -224,7 +224,8 @@ def __init__(self,
                  dilation: Union[int, Tuple[int, ...]] = 1,
                  groups: int = 1,
                  deform_groups: int = 1,
-                 bias: bool = False) -> None:
+                 bias: bool = False,
+                 im2col_step: int = 32) -> None:
```
Comment (MeowZheng) [inline mmcv/ops/deform_conv.py:228]: Add a definition for `im2col_step` in the docstring, and it would be better to provide solved issue link in the definition.

**Review by zhouzaida — APPROVED** [review]
LGTM

File: tests/test_ops/test_deform_conv.py [hunk 8]
```diff
@@ -45,9 +45,19 @@ def _test_deformconv(self,
...
+        x = torch.tensor(repeated_input, device=device, dtype=dtype)
         x.requires_grad = True
-        model = DeformConv2dPack(c_in, c_out, 2, stride=1, padding=0)
+        model = DeformConv2dPack(
+            in_channels=c_in,
+            out_channels=c_out,
+            kernel_size=2,
+            stride=1,
+            padding=0,
+            im2col_step=batch_size // 5)
```
Comment (MeowZheng) [inline tests/test_ops/test_deform_conv.py:60]: Shall we add two test cases: one tests `batch_size % im2col_step != 0` and the other tests  `batch_size < im2col_step`

Comment (teamwong111) [reply tests/test_ops/test_deform_conv.py]: yes, it's necessary. I have added.

File: mmcv/ops/deform_conv.py [hunk 9]
```diff
@@ -210,7 +210,10 @@ class DeformConv2d(nn.Module):
         deform_groups (int): Number of deformable group partitions.
         bias (bool): If True, adds a learnable bias to the output.
             Default: False.
-
+        im2col_step(int): Number of samples processed by the im2col_cuda_kernel
+        per call. If batch_size > im2col_step, batch_size % im2col_step == 0.
+        Solved from https://github.com/open-mmlab/mmcv/issues/985.
```
Comment (MeowZheng) [inline mmcv/ops/deform_conv.py:215]: It will work when batch_size > im2col_step, but batch_size  must be divisible by im2col_step. Solved this `issue <https://github.com/open-mmlab/mmcv/issues/985>`_.

**Review by MeowZheng — APPROVED** [review]

File: mmcv/ops/deform_conv.py [hunk 10]
```diff
@@ -210,7 +210,11 @@ class DeformConv2d(nn.Module):
         deform_groups (int): Number of deformable group partitions.
         bias (bool): If True, adds a learnable bias to the output.
             Default: False.
-
+        im2col_step(int): Number of samples processed by the im2col_cuda_kernel
+        per call. It will work when batch_size > im2col_step, but batch_size
+        must be divisible by im2col_step. Solved this issue
```
Comment (ZwwWayne) [inline mmcv/ops/deform_conv.py:215]: Indicate the next release version when this argument is added like `Added in v1.3.1x`

Comment (ZwwWayne) [reply mmcv/ops/deform_conv.py:215]: Better to add a note section to discuss the im2col_step and attach the issue link there.

File: mmcv/ops/deform_conv.py [hunk 11]
```diff
@@ -210,7 +210,11 @@ class DeformConv2d(nn.Module):
         deform_groups (int): Number of deformable group partitions.
         bias (bool): If True, adds a learnable bias to the output.
             Default: False.
-
+        im2col_step(int): Number of samples processed by the im2col_cuda_kernel
```
Comment (ZwwWayne) [inline mmcv/ops/deform_conv.py:213]: im2col_step(int) -> im2col_step (int)

File: tests/test_ops/test_deform_conv.py [hunk 12]
```diff
@@ -147,6 +182,12 @@ def test_deformconv(self):
         self._test_deformconv(torch.double)
         self._test_deformconv(torch.float)
         self._test_deformconv(torch.half, threshold=1e-1)
+        # test batch_size < im2col_step
+        self._test_deformconv(torch.float, batch_size=1, im2col_step=2)
+        # test bach_size % im2col_step != 0
+        with pytest.raises(
+                AssertionError, match='im2col step must divide batchsize'):
```
Comment (ZwwWayne) [inline tests/test_ops/test_deform_conv.py:189]: May also fix the error message like: `im2col_step must be divisible by batch size

File: mmcv/ops/deform_conv.py (hunk 10 above)
Comment (zhouzaida) [reply mmcv/ops/deform_conv.py:215]: Reference
https://github.com/open-mmlab/mmcv/blob/e22740b1d6953d75a0acecce4455d23800b1f018/mmcv/runner/hooks/evaluation.py#L64

`New in version 1.3.17.`

File: mmcv/ops/deform_conv.py [hunk 13]
```diff
@@ -210,11 +217,13 @@ class DeformConv2d(nn.Module):
...
-        per call. It will work when batch_size > im2col_step, but batch_size
-        must be divisible by im2col_step. Solved this issue
-        <https://github.com/open-mmlab/mmcv/issues/985>.
+        im2col_step (int): Number of samples processed by the
+        im2col_cuda_kernel per call. It will work when
+        batch_size > im2col_step, but batch_size must be divisible by
+        im2col_step. Solved this issue
+        <https://github.com/open-mmlab/mmcv/issues/1440>_.
             Default: 32.
+            `New in version 1.3.17.`
```
Comment (zhouzaida) [inline mmcv/ops/deform_conv.py]:
```suggestion
        im2col_step (int): Number of samples processed by the im2col_cuda_kernel
            per call. It will work when batch_size > im2col_step, but batch_size must be 
            divisible by im2col_step. Default: 32.
            `New in version 1.3.17.`
```

**Review by zhouzaida — APPROVED** [review]

File: mmcv/ops/deform_conv.py [hunk 14]
```diff
@@ -197,6 +197,13 @@ class DeformConv2d(nn.Module):
     `Deformable Convolutional Networks
     <https://arxiv.org/pdf/1703.06211.pdf>`_
 
+    Note:
+        The argument ``im2col_step`` was added in version 1.3.17, which means
+        number of samples processed by the ``im2col_cuda_kernel`` per call.
+        It enables users to define ``batch_size`` and ``im2col_step`` more
+        flexibly.Solved this `issue
```
Comment (ZwwWayne) [inline mmcv/ops/deform_conv.py:204]: It enables users to define ``batch_size`` and ``im2col_step`` more flexibly and solved `issue mmcv#1440 <xxx>`_

File: mmcv/ops/deform_conv.py [hunk 15]
```diff
@@ -210,7 +217,13 @@ class DeformConv2d(nn.Module):
         deform_groups (int): Number of deformable group partitions.
         bias (bool): If True, adds a learnable bias to the output.
             Default: False.
-
+        im2col_step (int): Number of samples processed by the
+            im2col_cuda_kernel per call. It will work when
+            batch_size > im2col_step, but batch_size must be divisible by
```
Comment (ZwwWayne) [inline mmcv/ops/deform_conv.py:222]: should we use '\`\`batch_size\`\`',  '\`\`im2col_step\`\`' here to indicate they are variables?

File: mmcv/ops/deform_conv.py [hunk 16]
```diff
@@ -210,7 +217,13 @@ class DeformConv2d(nn.Module):
...
+            im2col_step.Solved this `issue
+            <https://github.com/open-mmlab/mmcv/issues/1440>`_.
+            Default: 32.
+            `New in version 1.3.17.`
```
Comment (ZwwWayne) [inline mmcv/ops/deform_conv.py:226]: Make sure this style is consistent with the previous style.

**Review by ZwwWayne — APPROVED** [review]
